### PR TITLE
TOOLS-3575 Add coverage for mixed-schema timeseries buckets

### PR DESCRIPTION
**Pull request by tdq45gj**
This PR makes mongorestore handle `CannotInsertTimeseriesBucketsWithMixedSchema` by modifying the `timeseriesBucketsMayHaveMixedSchemaData` collection parameter. It adds a test to verify that mongodump can dump a timeseries bucket with mixed schema and another test to verify that restoring a bucket with mixed schema sets `timeseriesBucketsMayHaveMixedSchemaData` to false.

### CODE REVIEW

File: mongodump/mongodump_test.go
Comment (jddubois) [inline mongodump/mongodump_test.go:363]: It looks like backports were approved for v6.0 and v7.0. Can we create a ticket blocked on those backports to update these tests, and link a TODO here?

File: mongorestore/mongorestore_test.go
Comment (jddubois) [inline mongorestore/mongorestore_test.go:1679]: Same here

File: mongorestore/mongorestore_test.go
Comment (jddubois) [inline mongorestore/mongorestore_test.go:1714]: I'm a bit confused why this is succeeding- my understanding of the server change is that we need to set hasMixedSchema=true in order to perform these direct writes?

**Comment by tdq45gj** [comment]
@jddubois @autarch I made the changes to mongorestore but had to disable the test on v8.0 because of [SERVER-92222](https://jira.mongodb.org/browse/SERVER-92222). The server bug is on v8.0, which gives a wrong error code for the `CannotInsertTimeseriesBucketsWithMixedSchema` error.

File: common/db/db.go
Comment (jddubois) [inline common/db/db.go:565]: It looks like this is only checking the first error? Should we instead be returning true if any or all are 408 errors?

File: mongorestore/restore.go
Comment (jddubois) [inline mongorestore/restore.go:548]: Is it safe to call `ResetBulk()` twice here, or should we use `TryFlush`?

**Review by jddubois — COMMENTED** [review]
A few more comments- where did you disable the tests? Maybe we should just wait until the backport is merged? It looks like it's in code review.

File: common/db/db.go
Comment (tdq45gj) [reply common/db/db.go:565]: Good catch!

**Comment by tdq45gj** [comment]
@jddubois The test is disabled on v8.0 [here](https://github.com/mongodb/mongo-tools/pull/692/files#diff-57bd455e42f464e58586dd36657a3f51c0cbd32868d4679f119910874c964d68R1689-R1694). I think we need to wait for the bug fix to be backported to v8.0 and released. Given that the tools and server cyclic release dependency is still a problem, I don't think we can wait for that. I've tested this on v8.0 with its special error code (409) and it works.

File: mongorestore/mongorestore_test.go
Comment (jddubois) [inline mongorestore/mongorestore_test.go:1689]: Let's just add a TODO here with a linked REP ticket that's blocked on that backport? Or if not a TODO, as long as we have the info in the ticket that's fine.

**Review by jddubois — APPROVED** [review]
LGTM!

File: mongorestore/mongorestore_test.go
Comment (tdq45gj) [reply mongorestore/mongorestore_test.go:1689]: I added that info to [TOOLS-3597](https://jira.mongodb.org/browse/TOOLS-3597).

File: common/db/db.go
Comment (autarch) [inline common/db/db.go:569]: I think the logic here is still incorrect. If there are multiple write errors and the first one _isn't_ the one we check, but a later one is, then we'd return false incorrectly.

File: common/db/db.go
Comment (autarch) [inline common/db/db.go:599]: nit: We usually abbreviate "collection" to "coll", not "col".

File: mongorestore/restore.go
Comment (autarch) [inline mongorestore/restore.go:545]:
```suggestion
					resultChan <- result.withErr(errors.Wrap(collModErr, "failed to enable mixed schema in a timeseries bucket"))
```

But actually, I think this wrapping should be done in the `AllowMixedSchemaInTimeseriesBucket` method, rather than here.

**Review by autarch — CHANGES_REQUESTED** [review]
This looks good overall but there are a few things I think need to be updated.

File: common/db/db.go
Comment (tdq45gj) [reply common/db/db.go]: I changed the logic a bit for it to return true if any of the errors is the one we check.

File: common/db/db.go
Comment (autarch) [reply common/db/db.go]: Did you change this locally but not push it?

**Review by autarch — APPROVED** [review]
LGTM % one thing from the previous review. No need for another review.

File: common/db/db.go
Comment (tdq45gj) [reply common/db/db.go]: oops, I changed the parameter name but omitted the function name.

Comment (autarch) [reply common/db/db.go]: Also the variable _in_ the func. Just search for "colname".

**Pull request merged**
